powerstore/resource_volume.go

```diff
@@ -3,261 +3,334 @@ package powerstore
-	"github.com/hashicorp/terraform-plugin-framework/diag"
-	"github.com/hashicorp/terraform-plugin-framework/tfsdk"
-	"github.com/hashicorp/terraform-plugin-framework/types"
 	"log"
 	"strings"
+	client "terraform-provider-powerstore/client"
 	"terraform-provider-powerstore/models"
+
+	"github.com/dell/gopowerstore"
+	"github.com/hashicorp/terraform-plugin-framework-validators/stringvalidator"
+	"github.com/hashicorp/terraform-plugin-framework/path"
+	"github.com/hashicorp/terraform-plugin-framework/resource"
+	"github.com/hashicorp/terraform-plugin-framework/resource/schema"
+	"github.com/hashicorp/terraform-plugin-framework/resource/schema/planmodifier"
+	"github.com/hashicorp/terraform-plugin-framework/schema/validator"
+	"github.com/hashicorp/terraform-plugin-framework/types"
 )
 
 const defaultSectorSize = 512
 
-type resourceVolumeType struct{}
+type volumeResource struct {
+	client *client.Client
+}
 
-// GetSchema returns the schema for this resource.
-func (r resourceVolumeType) GetSchema(_ context.Context) (tfsdk.Schema, diag.Diagnostics) {
-	return tfsdk.Schema{
-		Attributes: map[string]tfsdk.Attribute{
-			"id": {
-				Type:                types.StringType,
-				Computed:            true,
+// Ensure the implementation satisfies the expected interfaces.
+var (
+	_ resource.Resource                = &volumeResource{}
+	_ resource.ResourceWithConfigure   = &volumeResource{}
+	_ resource.ResourceWithImportState = &volumeResource{}
+)
+
+// NewVolumeResource is a helper function to simplify the provider implementation.
+func newVolumeResource() resource.Resource {
+	return &volumeResource{}
+}
+
+func (r volumeResource) Metadata(ctx context.Context, req resource.MetadataRequest, resp *resource.MetadataResponse) {
+	resp.TypeName = req.ProviderTypeName + "_volume"
+}
+
+func (r volumeResource) Schema(ctx context.Context, req resource.SchemaRequest, resp *resource.SchemaResponse) {
+	resp.Schema = schema.Schema{
+		Description: ".",
+		Attributes: map[string]schema.Attribute{
+			"id": schema.StringAttribute{
 				Description:         "The ID of the volume.",
 				MarkdownDescription: "The ID of the volume.",
+				Computed:            true,
 			},
-			"name": {
-				Type:                types.StringType,
-				Required:            true,
+			"name": schema.StringAttribute{
 				Description:         "The name of the volume.",
 				MarkdownDescription: "The name of the volume.",
-			},
-			"size": {
-				Type:                types.Float64Type,
 				Required:            true,
+				Validators: []validator.String{
+					stringvalidator.LengthAtLeast(1),
+				},
+			},
+			"size": schema.Float64Attribute{
 				Description:         "The size of the volume.",
 				MarkdownDescription: "The size of the volume.",
+				Required:            true,
 			},
-			"capacity_unit": {
-				Type:                types.StringType,
-				Optional:            true,
-				Computed:            true,
+			"capacity_unit": schema.StringAttribute{
```

Comment:
can we specify the choices here?

Reply:
The check was present already but now choices are also added